corgi/core/models.py

```diff
@@ -421,17 +423,49 @@ def builds(self) -> QuerySet:
 
     @property
     def cpes(self) -> tuple[str, ...]:
-        """Return CPEs for all descendant variants."""
-        variant_cpes = (
+        """Return CPEs for direct descendant variants if they exist. Otherwise, return a union
+        of indirect descendant variants and descendant streams cpes_matching_patterns"""
+        if self is ProductVariant:
+            return (self.cpe,)  # type: ignore[attr-defined]
+
+        direct_variant_cpes = (
             self.pnodes.get_queryset()
             .get_descendants(include_self=True)
             .using("read_only")
             .filter(level=MODEL_NODE_LEVEL_MAPPING["ProductVariant"])
+            .exclude(type=ProductNode.ProductNodeType.INFERRED)
             .values_list("productvariant__cpe", flat=True)
             .distinct()
         )
-        # Omit CPEs like "", but GenericForeignKeys only support .filter(), not .exclude()??
-        return tuple(cpe for cpe in variant_cpes if cpe)
+        if direct_variant_cpes:
+            return self.qs_to_tuple_no_empty_strings(direct_variant_cpes)
+
+        cpes_matching_patterns = (
+            self.pnodes.get_queryset()
+            .get_descendants(include_self=True)
+            .using("read_only")
+            .filter(level=MODEL_NODE_LEVEL_MAPPING["ProductStream"])
+            .values_list("productstream__cpes_matching_patterns", flat=True)
+            .distinct()
+        )
+        indirect_variant_cpes = (
+            self.pnodes.get_queryset()
+            .get_descendants(include_self=True)
+            .using("read_only")
+            .filter(
+                level=MODEL_NODE_LEVEL_MAPPING["ProductVariant"],
+                type=ProductNode.ProductNodeType.INFERRED,
+            )
+            .values_list("productvariant__cpe", flat=True)
+            .distinct()
+        )
+        distinct_cpes_matching_patterns = set([cpe for cpe_patterns in cpes_matching_patterns for cpe in cpe_patterns])
+        distinct_cpes_matching_patterns.update(self.qs_to_tuple_no_empty_strings(indirect_variant_cpes))
+        return tuple(distinct_cpes_matching_patterns)
+
+    def qs_to_tuple_no_empty_strings(self, direct_variant_cpes: QuerySet) -> tuple[str, ...]:
```

Comment:
If you change above, you can use Iterable for the type hint instead:
```suggestion
    def qs_to_tuple_no_empty_strings(self, direct_variant_cpes: Iterable[str]) -> tuple[str, ...]:
```